src/editconfig.cpp

```diff
@@ -0,0 +1,559 @@
+				};
+
+				connect( pItemButton, &QPushButton::pressed, this, onItemButtonPushedCallback );
+			}
+
+			// We now unravel our stored contents, a QPair containing a QString
+			// and QJsonArray pointer. We grab the string for the header's name
+			// and the QJsonArray for the header's contents. We need to
+			// convert it from a pointer into a reference first.
+			for ( const auto &jsonArrayStringPair : jsonArrayVector )
+			{
+				qInfo() << jsonArrayStringPair.first;
+				auto headerObject = QJsonObject();
+				headerObject["header"] = jsonArrayStringPair.first;
+				headerObject["content"] = *jsonArrayStringPair.second;
+				mainJSONArray.append( headerObject );
+				// we delete the QJsonArray pointers, I don't trust QT enough to get rid of them after
+				// QVector is done with them.
+				delete jsonArrayStringPair.second;
+			}
+			configJSONDocument.setArray( mainJSONArray );
+			auto configFile = QFile( "./config.json" );
+			configFile.open( QFile::WriteOnly );
+			configFile.write( configJSONDocument.toJson() );
+			configFile.close();
+		}
+		this->close();
+	};
+
+	// We then use these callbacks.
+	connect( m_pEditList, &QListWidget::currentRowChanged, this, onCurrentRowChangedCallback );
+	connect( addCurrentButton, &QPushButton::pressed, this, onAddItemButtonPressedCallback );
+	connect( shiftDownButton, &QPushButton::pressed, this, onShiftItemDownButtonPressedCallback );
+	connect( editButton, &QPushButton::pressed, this, onEditItemButtonPressedCallback );
+	connect( shiftUpButton, &QPushButton::pressed, this, onShiftItemUpButtonPressedCallback );
+	connect( removeCurrentButton, &QPushButton::pressed, this, onRemoveItemButtonPressedCallback );
+	connect( dialogButtons, &QDialogButtonBox::clicked, this, onDialogButtonsPressedCallback );
+
+	// to prevent buttons from aligning equally across the m_pEditList widget's
+	// height, we cap the stretch at 1, keeping them all neatly lined up.
+	pEditConfigurationLayout->setRowStretch( 4, 1 );
+
+	// Set focus, so we don't have focus directly on the top most item
+	this->setFocus( Qt::NoFocusReason );
+}
+
+ui::CEditConfigPopup::CEditConfigPopup( CEditConfig *parent ) :
+	QDialog( parent )
+{
+	auto pConfigPopupLayout = new QGridLayout( this );
+	pConfigPopupLayout->setAlignment( Qt::AlignTop | Qt::AlignRight );
+
+	m_pNameLabel = new QLabel( this );
+	m_pNameLabel->setText( "Name:" );
+	pConfigPopupLayout->addWidget( m_pNameLabel, 0, 0 );
+	m_pNameLineEdit = new QLineEdit( this );
+	pConfigPopupLayout->addWidget( m_pNameLineEdit, 0, 1 );
+
+	m_pTypeLabel = new QLabel( this );
+	m_pTypeLabel->setText( "Type:" );
+	pConfigPopupLayout->addWidget( m_pTypeLabel, 0, 2 );
+	m_pTypeComboBox = new QComboBox( this );
+	m_pTypeComboBox->addItem( "Executable" );
+	m_pTypeComboBox->addItem( "Url" );
+	m_pTypeComboBox->addItem( "Category" );
+	pConfigPopupLayout->addWidget( m_pTypeComboBox, 0, 3 );
+
+	m_pUrlLabel = new QLabel( this );
```

Comment:
These don't really need to be members. You're not accessing them from anywhere outside this function?

Reply:
There was a weird bug early in development that required them to be, it had to do with callback lambdas not accessing them, but I figured out the cause, yet... forgot about these... I'll fix it.